In [6]:
import requests
import io
import pandas as pd
import regex as re

from bs4 import BeautifulSoup
from pdfminer.high_level import extract_text

from unpy_get_dates import *
from scihub_upgraded import SciHub

from datetime import datetime

import warnings
warnings.filterwarnings("ignore")

In [2]:
sh = SciHub()

In [3]:
#Successful DOI
sh.get_dates("10.1016/j.joi.2013.09.001")

['https://moscow.sci-hub.st/2341/219df39bc7707c05bbd4f1755fd9d3d6/10.1016@j.joi.2013.09.001.pdf#navpanes=0&view=FitH',
 '20  July  2013',
 '4  September  2013']

In [4]:
#No Date DOI
sh.get_dates("10.1111/aman.13399")

['http://sci-hub.st/downloads/2020-06-28//b0/10.1111@aman.13399.pdf#navpanes=0&view=FitH',
 'no_date_found']

In [5]:
# Successful but 10MB DOI
sh.get_dates("10.2138/am-2017-6003")

['https://twin.sci-hub.st/6374/664526fc806ec4a5039f11e7f00dbe16/jin2017.pdf#navpanes=0&view=FitH',
 'noveMber 2, 2016',
 'February 14, 2017']

In [14]:
%store -r Scihub_df


In [8]:
Scihub_df.Scihub_results.map(type).value_counts()

<class 'float'>    26691
<class 'list'>     19052
<class 'str'>        948
Name: Scihub_results, dtype: int64

As the whole dataset is taking too long, will seperate it to 10k chunks

## First Loop (for 10K):

In [4]:
for k in range(10):

    sh = SciHub()

    start = k*1000
    end = (k+1)*1000 
    doi_slice = Scihub_df[start:end].index.tolist()

    now = datetime.now()
    print(f"Starting loop {k+1}, start time: {now}")
    Scihub_df.loc[doi_slice,"Scihub_results"] = Scihub_df.loc[doi_slice,"doi"].map(sh.get_dates)
    now = datetime.now()
    print(f"{k+1} loop ended, endtime: {now}")
   

Starting loop 1, start time: 2022-01-05 13:01:48.165611


1 loop ended, endtime: 2022-01-05 14:32:21.898530
Starting loop 2, start time: 2022-01-05 14:32:21.899030


2 loop ended, endtime: 2022-01-05 15:54:50.545649
Starting loop 3, start time: 2022-01-05 15:54:50.546154


3 loop ended, endtime: 2022-01-05 17:11:31.362682
Starting loop 4, start time: 2022-01-05 17:11:31.363182


C:\Users\oguzk\anaconda3\lib\site-packages\pdfminer\pdfpage.py:159: PDFTextExtractionNotAllowedWarning: The PDF <_io.BytesIO object at 0x00000209DC723450> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding.
  warnings.warn(warning_msg, PDFTextExtractionNotAllowedWarning)
C:\Users\oguzk\anaconda3\lib\site-packages\pdfminer\pdfpage.py:159: PDFTextExtractionNotAllowedWarning: The PDF <_io.BytesIO object at 0x00000209DB95B7C0> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding.
  warnings.warn(warning_msg, PDFTextExtractionNotAllowedWarning)
C:\Users\oguzk\anaconda3\lib\site-packages\pdfminer\pdfpage.py:159: PDFTextExtractionNotAllowedWarning: The PDF <_io.BytesIO object at 0x00000209DBC56CC0> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding.
  warnings.warn(warning_msg, PDFTextExtractionNotAllowedWarn

4 loop ended, endtime: 2022-01-05 18:32:27.902407
Starting loop 5, start time: 2022-01-05 18:32:27.902407


5 loop ended, endtime: 2022-01-05 19:51:44.526113
Starting loop 6, start time: 2022-01-05 19:51:44.526113


6 loop ended, endtime: 2022-01-05 21:08:40.420269
Starting loop 7, start time: 2022-01-05 21:08:40.420845


7 loop ended, endtime: 2022-01-05 22:26:20.042316
Starting loop 8, start time: 2022-01-05 22:26:20.042316


8 loop ended, endtime: 2022-01-05 23:48:06.437624
Starting loop 9, start time: 2022-01-05 23:48:06.438124


9 loop ended, endtime: 2022-01-06 01:11:46.376742
Starting loop 10, start time: 2022-01-06 01:11:46.376742


10 loop ended, endtime: 2022-01-06 02:27:40.314846


### Analysis of First Loop:

In [45]:
Scihub_df.Scihub_results.map(type).value_counts()

<class 'float'>    36691
<class 'str'>       6108
<class 'list'>      3892
Name: Scihub_results, dtype: int64

In [47]:
Scihub_df[Scihub_df.Scihub_results.map(type) == str]["Scihub_results"].value_counts()

cant_open_pdf            5584
article_not_in Scihub     524
Name: Scihub_results, dtype: int64

The first 10k loop is finished, however 5584 of the results are "cant_open_pdf


In [48]:
Scihub_df[Scihub_df.Scihub_results.map(type) == list]["Scihub_results"].map(len).value_counts()

2    2038
3    1854
Name: Scihub_results, dtype: int64

In [5]:
Scihub_df[Scihub_df.Scihub_results.map(type) == list].loc[0,"Scihub_results"]

['https://zero.sci-hub.ru/29/c838ad3b574e083a56c278b8f786cb53/yan2010.pdf#navpanes=0&view=FitH',
 'no_date_found']

In [42]:
Scihub_df[Scihub_df.Scihub_results == "cant_open_pdf"].head()

doi  \
38              10.2308/accr-51296   
185  10.1080/00020184.2020.1755828   
233       10.1108/afr-11-2019-0120   
260             10.1111/aman.13399   
370             10.1093/aje/kwz107   

                                           doi_url  \
38              https://doi.org/10.2308/accr-51296   
185  https://doi.org/10.1080/00020184.2020.1755828   
233       https://doi.org/10.1108/afr-11-2019-0120   
260             https://doi.org/10.1111/aman.13399   
370             https://doi.org/10.1093/aje/kwz107   

                                                 title            genre  \
38   Mandatory Financial Reporting and Voluntary Di...  journal-article   
185  The proliferation of overlapping sub-regional ...  journal-article   
233  Design of the rainfall index annual forage pro...  journal-article   
260                              The Predatory Present  journal-article   
370  The Impact of 3 Years of Targeted Indoor Resid...  journal-article   

    published_date  year                      journal_name  \
38      2015-09-01  2015             The Accounting Review   
185     2020-01-02  2020                   African Studies   
233     2020-07-10  2020       Agricultural Finance Review   
260     2020-06-01  2020           American Anthropologist   
370     2019-05-07  2019  American Journal of Epidemiology   

           journal_issns journal_issn_l  journal_is_oa  journal_is_in_doaj  \
38   1558-7967,0001-4826      0001-4826          False               False   
185  0002-0184,1469-2872      0002-0184          False               False   
233            0002-1466      0002-1466          False               False   
260  0002-7294,1548-1433      0002-7294          False               False   
370  0002-9262,1476-6256      0002-9262          False               False   

                           publisher  is_oa oa_status  \
38   American Accounting Association   True     green   
185               Informa UK Limited  False    closed   
233                          Emerald  False    closed   
260                            Wiley  False    closed   
370    Oxford University Press (OUP)   True     green   

                                          oa_locations  \
38   [{'updated': '2021-08-22T11:52:53.217319', 'ur...   
185                                                 []   
233                                                 []   
260                                                 []   
370  [{'updated': '2021-12-15T13:25:14.509602', 'ur...   

                        updated  \
38   2021-01-20T18:25:56.939118   
185  2021-01-18T07:32:01.616343   
233  2021-02-14T01:36:33.643281   
260  2021-01-19T11:03:17.758547   
370  2021-03-28T21:15:50.062251   

                                             z_authors Unpy_filter  \
38   [{'given': 'Xi', 'family': 'Li', 'sequence': '...      Scihub   
185  [{'ORCID': 'http://orcid.org/0000-0003-4404-47...      Scihub   
233  [{'given': 'Abby', 'family': 'ShalekBriski', '...      Scihub   
260  [{'given': 'Alex W.', 'family': 'Barker', 'seq...      Scihub   
370  [{'ORCID': 'http://orcid.org/0000-0003-2471-67...      Scihub   

    Scihub_results  
38   cant_open_pdf  
185  cant_open_pdf  
233  cant_open_pdf  
260  cant_open_pdf  
370  cant_open_pdf

However, when checked again, the function is actually able to extract the date information correctly as can be seen below. So, for Scihub, we have a problem, possibly CAPTCHA, that escapes all the checks and returns cant_open_pdf incorrectly. This must be solved somehow :(

In [11]:
sh = SciHub()

sh.get_dates("10.2308/accr-51296")

['http://sci-hub.ru/downloads/2019-08-24/b5/li2016.pdf#navpanes=0&view=FitH',
 'February 2014',
 'July 2015']

Storing the dataset so far, but not starting the second loop

In [49]:
%store Scihub_df

Stored 'Scihub_df' (DataFrame)


In [ ]:

for k in range(10,20):
    
    start = k*1000
    end = (k+1)*1000 
    doi_slice = Scihub_df[start:end].index.tolist()

    now = datetime.now()
    print(f"Starting loop {k+1}, start time: {now}")
    Scihub_df.loc[doi_slice,"Scihub_results"] = Scihub_df.loc[doi_slice,"doi"].map(sh.get_dates)
    now = datetime.now()
    print(f"{k+1} loop ended, endtime: {now}")

    

In [23]:
pd.cut(Scihub_df[Scihub_df.Scihub_results == "cant_open_pdf"].index, bins=[0,1000,2000,3000,4000,5000,6000,7000,8000,9000,10000]).value_counts()

(0, 1000]         10
(1000, 2000]     143
(2000, 3000]     627
(3000, 4000]     561
(4000, 5000]     644
(5000, 6000]     591
(6000, 7000]     684
(7000, 8000]     634
(8000, 9000]     621
(9000, 10000]    119
dtype: int64

## Second Loop (10K - 15K )

In [5]:
for k in range(5):

    sh = SciHub()
    
    start = (k+10)*1000
    end = (k+11)*1000 
    doi_slice = Scihub_df[start:end].index.tolist()

    now = datetime.now()
    print(f"Starting loop {k+1}, start time: {now}")
    Scihub_df.loc[doi_slice,"Scihub_results"] = Scihub_df.loc[doi_slice,"doi"].map(sh.get_dates)
    now = datetime.now()
    print(f"{k+1} loop ended, endtime: {now}")
   

Starting loop 1, start time: 2022-01-26 15:34:46.320698
1 loop ended, endtime: 2022-01-26 16:20:04.340549
Starting loop 2, start time: 2022-01-26 16:20:04.342550
2 loop ended, endtime: 2022-01-26 17:04:08.846540
Starting loop 3, start time: 2022-01-26 17:04:08.848540
3 loop ended, endtime: 2022-01-26 17:54:07.392152
Starting loop 4, start time: 2022-01-26 17:54:07.394654
4 loop ended, endtime: 2022-01-26 18:35:31.117726
Starting loop 5, start time: 2022-01-26 18:35:31.119727
5 loop ended, endtime: 2022-01-26 19:16:41.131252


In [28]:
%store Scihub_df

Stored 'Scihub_df' (DataFrame)


### Analysis of Second Loop:

In [7]:
Scihub_df[10000:15000]["Scihub_results"].map(type).value_counts()

<class 'list'>    4731
<class 'str'>      269
Name: Scihub_results, dtype: int64

At a first glance, the second loop looks very promising and got much better results than the first loop with very little type str returns. 

In [26]:
Scihub_df[10000:15000][Scihub_df[10000:15000]["Scihub_results"].map(type)==str]["Scihub_results"].value_counts()

article_not_in Scihub    267
cant_open_pdf              2
Name: Scihub_results, dtype: int64

When the errors are analysed, it is seen that all except 2 were actually unavailable articles. However, 2 error articles are actually available on Scihub. So ?

In [27]:
Scihub_df[10000:15000][Scihub_df[10000:15000]["Scihub_results"]=="cant_open_pdf"]

doi                            doi_url  \
12863   10.2307/25750704   https://doi.org/10.2307/25750704   
12880  10.1002/pam.21919  https://doi.org/10.1002/pam.21919   

                                                   title            genre  \
12863  Improving Employees' Compliance Through Inform...  journal-article   
12880  Energy-Based Economic Development: How Clean E...  journal-article   

      published_date  year                               journal_name  \
12863     2010-01-01  2010                              MIS Quarterly   
12880     2016-06-01  2016  Journal of Policy Analysis and Management   

      journal_issns journal_issn_l  journal_is_oa  journal_is_in_doaj  \
12863     0276-7783      0276-7783          False               False   
12880     0276-8739      0276-8739          False               False   

      publisher  is_oa oa_status oa_locations                     updated  \
12863     JSTOR  False    closed           []  2021-04-26T09:42:09.304629   
12880     Wiley  False    closed           []  2021-01-13T15:02:12.369641   

                                               z_authors Unpy_filter  \
12863  [{'family': 'Puhakainen', 'sequence': 'first'}...      Scihub   
12880  [{'given': 'Karnamadakala Rahul', 'family': 'S...      Scihub   

      Scihub_results  
12863  cant_open_pdf  
12880  cant_open_pdf

In [11]:
Scihub_df[10000:15000][Scihub_df[10000:15000]["Scihub_results"].map(type)==list]["Scihub_results"].map(len).value_counts()

3    2519
2    2212
Name: Scihub_results, dtype: int64

When the successful (list) returns are analysed, it's seen that more than 50% have at least 1 date value, which is simply amazing

In [25]:
Scihub_df[10000:15000][(Scihub_df[10000:15000]["Scihub_results"].map(type)==list) & (Scihub_df[10000:15000]["Scihub_results"].map(len)==3)]["Scihub_results"].map(lambda x: x[0]).map(type).value_counts()

<class 'str'>    2519
Name: Scihub_results, dtype: int64

In [24]:
Scihub_df[10000:15000][(Scihub_df[10000:15000]["Scihub_results"].map(type)==list) & (Scihub_df[10000:15000]["Scihub_results"].map(len)==3)]["Scihub_results"].map(lambda x: x[1]).map(type).value_counts()

<class 'str'>         2319
<class 'NoneType'>     200
Name: Scihub_results, dtype: int64

Out of 2519, only 200 does not have "Received Date" information

In [23]:
Scihub_df[10000:15000][(Scihub_df[10000:15000]["Scihub_results"].map(type)==list) & (Scihub_df[10000:15000]["Scihub_results"].map(len)==3)]["Scihub_results"].map(lambda x: x[2]).map(type).value_counts()

<class 'str'>         2230
<class 'NoneType'>     289
Name: Scihub_results, dtype: int64

Out of 2519, only 289 does not have "Accepted Date" information

## Re-Run First Loop

As the upgrades & improvements had quite an impact on the process, it would be beneficial to re-run the first loop and compare the two results

In [29]:
first_run_df = Scihub_df[0:10000].copy()

In [30]:
%store first_run_df

Stored 'first_run_df' (DataFrame)


In [12]:
for k in range(10):

    sh = SciHub()

    start = k*1000
    end = (k+1)*1000 
    doi_slice = Scihub_df[start:end].index.tolist()

    now = datetime.now()
    print(f"Starting loop {k+1}, start time: {now}")
    Scihub_df.loc[doi_slice,"Scihub_results"] = Scihub_df.loc[doi_slice,"doi"].map(sh.get_dates)
    now = datetime.now()
    print(f"{k+1} loop ended, endtime: {now}")
   

Starting loop 1, start time: 2022-01-27 19:42:27.831609


ConnectionError: ('Connection aborted.', TimeoutError(10060, 'A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond', None, 10060, None))

While re-running the first loop, we have encountered a new type of error. After some debugging, it is seen that the main reason for the error is that in rare cases, Sci-Hub does not use an "iframe" attribute but rather "embed". So, there is no ["src"] field in the soup, causing an error, which is only dealt with during the last "Unknown Error" check. 

However, as there was another issue with change_base_url, @retry causes the base_url to be None, which stopped the entire loop in loop 3. The results for the 1-2 loops are successfully saved.

As a result, the scihub_upgraded was updated to fix these 2 issues.

But, 2000-10000 loops should be run again!

In [140]:
%store Scihub_df

Stored 'Scihub_df' (DataFrame)


## Re-Run First Loop (2K - 3K)

All the issues in "re-run first loop" is (hopefully) fixed:

    1- for loop with ["iframe" ,"embed"]

    2- _change_base_url works succesfully

    3- zero.sci-hub -> moscow.sci-hub

Will run the problematic third loop (2K - 3K) , if executes succesfully, will run the rest.

In [14]:
sh = SciHub()

doi_slice = Scihub_df[2000:3000].index.tolist()

now = datetime.now()
print(f"Starting loop, start time: {now}")
Scihub_df.loc[doi_slice,"Scihub_results"] = Scihub_df.loc[doi_slice,"doi"].map(sh.get_dates)
now = datetime.now()
print(f"loop ended, endtime: {now}")
   

Starting loop, start time: 2022-01-28 10:31:51.083441


loop ended, endtime: 2022-01-28 12:37:47.713046


In [8]:
Scihub_df[2000:3000]["Scihub_results"].map(type).value_counts()

<class 'list'>     938
<class 'str'>       51
<class 'tuple'>     11
Name: Scihub_results, dtype: int64

In [9]:
Scihub_df[2000:3000][Scihub_df[2000:3000]["Scihub_results"].map(type)==list]["Scihub_results"].map(len).value_counts()

3    497
2    441
Name: Scihub_results, dtype: int64

In [10]:
Scihub_df[2000:3000][Scihub_df[2000:3000]["Scihub_results"].map(type)==str]["Scihub_results"].value_counts()

article_not_in Scihub    51
Name: Scihub_results, dtype: int64

In [11]:
Scihub_df[2000:3000][Scihub_df[2000:3000]["Scihub_results"].map(type)==tuple]["Scihub_results"]

2576    (https://moscow.sci-hub.st/3241/51d11ee361f5a5...
2657    (https://moscow.sci-hub.se/3316/351eab13f521e4...
2659    (https://moscow.sci-hub.se/5461/41b5ac3322b902...
2687    (https://moscow.sci-hub.se/3790/454c9651e91c34...
2766    (https://moscow.sci-hub.se/1532/d218032da8973b...
2791    (https://moscow.sci-hub.se/2097/fb20c00f76352e...
2802    (https://moscow.sci-hub.se/2055/e5b198b4efec35...
2824    (https://moscow.sci-hub.se/355/0c2bf696b294848...
2828    (https://moscow.sci-hub.se/2575/aa39429cbd16d0...
3283    (http://index.ggws.net/downloads/2021-05-26/21...
3647    (http://index.ggws.net/downloads/2020-02-04/d5...
Name: Scihub_results, dtype: object

FINALLY, it seems that loop is executed succesfully. 

Although it took longer than expected (125 mins), everything seems to be in order. 

Only, small problem is that, as we are replacing zero. domains with moscow. , this caused 11 tuple results with "cant open pdf" returns. This is fixed now in scihub_upgraded.

NOTE & UPDATE: It seems that the Zero-Moscow error is not due to an error in the code but rather due to the SciHub servers. So this fix is reverted and zero.sci-ub domains are not altered in any way. 

The cell below aims to fix these 11 tuple results, if this also works, the re-run for 3K - 10K can be run too.



In [13]:
Scihub_df.loc[Scihub_df.Scihub_results.map(type)==tuple, "Scihub_results"] = Scihub_df.loc[Scihub_df.Scihub_results.map(type)==tuple, "doi"].map(sh.get_dates)

In [11]:
Scihub_df[2000:3000]["Scihub_results"].map(type).value_counts()

<class 'list'>    948
<class 'str'>      52
Name: Scihub_results, dtype: int64

In [19]:
Scihub_df.loc[2828,"Scihub_results"]

['https://sci.bban.top/pdf/10.1016/j.indmarman.2014.05.012.pdf#view=FitH',
 '15 February 2014',
 '30 March 2014']

In [29]:
Scihub_df.to_csv("0-3and10-15done_Scihub_df.csv")

In [13]:
Scihub_df[:3000]["Scihub_results"].map(type).value_counts()

<class 'list'>    2818
<class 'str'>      182
Name: Scihub_results, dtype: int64

In [12]:
Scihub_df[:3000][Scihub_df[:3000]["Scihub_results"].map(type)==list]["Scihub_results"].map(len).value_counts()

2    1422
3    1396
Name: Scihub_results, dtype: int64

In [15]:
Scihub_df[10000:15000][Scihub_df[10000:15000]["Scihub_results"].map(type)==list]["Scihub_results"].map(len).value_counts()

3    2519
2    2212
Name: Scihub_results, dtype: int64

## Re-Run First Loop (Take Three) (3K - 10K)

In [13]:
for k in range(3,10):


    start = k*1000
    end = (k+1)*1000 
    doi_slice = Scihub_df[start:end].index.tolist()

    now = datetime.now()
    print(f"Starting loop {k+1}, start time: {now}")
    Scihub_df.loc[doi_slice,"Scihub_results"] = Scihub_df.loc[doi_slice,"doi"].map(sh.get_dates)
    now = datetime.now()
    print(f"{k+1} loop ended, endtime: {now}")

    sh = SciHub()
    
   

Starting loop 4, start time: 2022-01-28 15:00:32.482995


4 loop ended, endtime: 2022-01-28 16:03:47.718905
Starting loop 5, start time: 2022-01-28 16:03:47.721405


5 loop ended, endtime: 2022-01-28 17:19:00.455010
Starting loop 6, start time: 2022-01-28 17:19:00.460505


6 loop ended, endtime: 2022-01-28 18:19:27.182961
Starting loop 7, start time: 2022-01-28 18:19:27.187960


7 loop ended, endtime: 2022-01-28 19:34:40.265538
Starting loop 8, start time: 2022-01-28 19:34:40.270039


8 loop ended, endtime: 2022-01-28 20:33:13.027141
Starting loop 9, start time: 2022-01-28 20:33:13.029641


9 loop ended, endtime: 2022-01-28 21:42:14.315251
Starting loop 10, start time: 2022-01-28 21:42:14.317252


10 loop ended, endtime: 2022-01-28 22:42:54.297158


In [15]:
Scihub_df[3000:10000]["Scihub_results"].map(type).value_counts()

<class 'list'>    6592
<class 'str'>      408
Name: Scihub_results, dtype: int64

In [17]:
Scihub_df[3000:10000][Scihub_df[3000:10000]["Scihub_results"].map(type)==str]["Scihub_results"].value_counts()

article_not_in Scihub          350
cant_read_pdf (Zero-Moscow)     58
Name: Scihub_results, dtype: int64

In [23]:
Scihub_df[3000:10000][Scihub_df[3000:10000]["Scihub_results"].map(type)==list]["Scihub_results"].map(len).value_counts()

2    3320
3    3272
Name: Scihub_results, dtype: int64

In [22]:
Scihub_df[3000:10000][Scihub_df[3000:10000]["Scihub_results"]=="article_not_in Scihub"]

doi  \
3869            10.1162/jinh_r_01221   
3902          10.18647/2988/jjs-2010   
3903          10.18647/3022/jjs-2011   
3904          10.18647/3089/jjs-2012   
3905          10.18647/3185/jjs-2014   
...                              ...   
11872         10.3366/para.2015.0158   
11873         10.3366/para.2016.0203   
11897   10.1097/eja.0000000000001320   
11925  10.1108/imr.2010.03627aaa.001   
12010      10.1017/s0266078420000115   

                                             doi_url  \
3869            https://doi.org/10.1162/jinh_r_01221   
3902          https://doi.org/10.18647/2988/jjs-2010   
3903          https://doi.org/10.18647/3022/jjs-2011   
3904          https://doi.org/10.18647/3089/jjs-2012   
3905          https://doi.org/10.18647/3185/jjs-2014   
...                                              ...   
11872         https://doi.org/10.3366/para.2015.0158   
11873         https://doi.org/10.3366/para.2016.0203   
11897   https://doi.org/10.1097/eja.0000000000001320   
11925  https://doi.org/10.1108/imr.2010.03627aaa.001   
12010      https://doi.org/10.1017/s0266078420000115   

                                                   title            genre  \
3869   Jihād in West Africa during the Age of Revolut...  journal-article   
3902   Il tempio di Leontopoli in Egitto: Identità po...  journal-article   
3903   Conceiving Israel: The Fetus in Rabbinic Narra...  journal-article   
3904                      Mémoires des Juifs de Roumanie  journal-article   
3905   And I shall dwell in their midst: God’s presen...  journal-article   
...                                                  ...              ...   
11872  A Vocabulary and Its Vicissitudes: Notes towar...  journal-article   
11873  The Uses of Maurice Blanchot in Bernard Stiegl...  journal-article   
11897  The ‘ephemeral’ intensive care units that save...  journal-article   
11925                                          Editorial  journal-article   
12010  English spelling: where do -tion and -sion com...  journal-article   

      published_date  year                              journal_name  \
3869      2018-02-01  2018  The Journal of Interdisciplinary History   
3902      2010-10-01  2010                 Journal of Jewish Studies   
3903      2011-04-01  2011                 Journal of Jewish Studies   
3904      2012-04-01  2012                 Journal of Jewish Studies   
3905      2014-10-01  2014                 Journal of Jewish Studies   
...              ...   ...                                       ...   
11872     2015-07-01  2015                                 Paragraph   
11873     2016-11-01  2016                                 Paragraph   
11897     2020-08-26  2020       European Journal of Anaesthesiology   
11925     2010-02-23  2010            International Marketing Review   
12010     2020-07-01  2020                             English Today   

             journal_issns journal_issn_l  journal_is_oa  journal_is_in_doaj  \
3869   0022-1953,1530-9169      0022-1953          False               False   
3902   0022-2097,2056-6689      0022-2097          False               False   
3903   0022-2097,2056-6689      0022-2097          False               False   
3904   0022-2097,2056-6689      0022-2097          False               False   
3905   0022-2097,2056-6689      0022-2097          False               False   
...                    ...            ...            ...                 ...   
11872  0264-8334,1750-0176      0264-8334          False               False   
11873  0264-8334,1750-0176      0264-8334          False               False   
11897  0265-0215,1365-2346      0265-0215          False               False   
11925            0265-1335      0265-1335          False               False   
12010  0266-0784,1474-0567      0266-0784          False               False   

                                       publisher  is_oa oa_status  \
3869                        MIT Press - Journals  False    

In [29]:
Scihub_df[3000:10000][Scihub_df[3000:10000]["Scihub_results"]=="cant_read_pdf (Zero-Moscow)"]

doi  \
5040         10.1080/00288306.2010.500321   
5041         10.1080/00288306.2013.772906   
5043        10.1080/00288306.2015.1077872   
5047        10.1080/00293652.2015.1039569   
5052                10.1093/notesj/gjq125   
5698         10.1097/psy.0000000000000377   
5995                   10.1111/rssc.12417   
6140   10.1053/j.seminhematol.2020.05.001   
6154                    10.1093/sq/quz014   
6287        10.1016/s0038-1101(13)00240-2   
6288            10.1016/j.sse.2014.07.004   
6290             10.1215/00382876-1162597   
6292             10.1215/00382876-2345270   
6300     10.1111/j.1540-6237.2010.00686.x   
6301     10.1111/j.1540-6237.2011.00776.x   
6303                   10.1111/ssqu.12006   
6310               10.1002/star.201000005   
6314     10.1111/j.1467-9582.2012.01187.x   
7037                10.1353/sub.2012.0010   
7038                10.1353/sub.2014.0026   
7287            10.1017/s0075426920000300   
7408            10.1017/s0080440114000073   
7412     10.1111/j.1467-9477.2010.00250.x   
7415              10.1111/1467-9477.12028   
7432            10.1484/j.viator.1.103476   
7433                    10.1553/wst123s95   
7435     10.1163/008467210x12626615185621   
7437            10.1163/15736121-12341237   
7439            10.1163/15736121-12341292   
8651       10.1016/j.ijadhadh.2011.06.002   
8653       10.1016/j.ijadhadh.2013.04.006   
8658       10.1016/j.ijadhadh.2018.02.031   
8661            10.1108/01437731111169997   
8666        10.1016/s0143-8166(11)00086-8   
8669      10.1016/j.optlaseng.2013.11.007   
8670      10.1016/j.optlaseng.2014.05.006   
8671      10.1016/j.optlaseng.2015.11.009   
9105         10.1097/sap.0000000000000959   
9825          10.1016/j.physd.2010.08.007   
9826          10.1016/j.physd.2011.09.016   
9828          10.1016/j.physd.2013.12.001   
9830          10.1016/j.physd.2016.03.004   
9836           10.1016/j.cose.2011.05.002   
9837           10.1016/j.cose.2012.07.008   
9838           10.1016/j.cose.2013.08.006   
9846       10.1016/j.strusafe.2010.03.006   
9847       10.1016/j.strusafe.2011.04.003   
9849       10.1016/j.strusafe.2013.02.002   
9850       10.1016/j.strusafe.2013.08.003   
9851       10.1016/j.strusafe.2014.07.001   
10959         10.1016/j.appet.2009.10.004   
10960         10.1016/j.appet.2011.05.289   
10961         10.1016/j.appet.2012.05.059   
10962         10.1016/j.appet.2013.02.001   
10963         10.1016/j.appet.2014.07.015   
10970        10.1097/paf.0b013e3181dfc98b   
10973          10.1016/j.eiar.2011.04.003   
10975          10.1016/j.eiar.2014.05.004   

                                                 doi_url  \
5040        https://doi.org/10.1080/00288306.2010.500321   
5041        https://doi.org/10.1080/00288306.2013.772906   
5043       https://doi.org/10.1080/00288306.2015.1077872   
5047       https://doi.org/10.1080/00293652.2015.1039569   
5052               https://doi.org/10.1093/notesj/gjq125   
5698        https://doi.org/10.1097/psy.0000000000000377   
5995                  https://doi.org/10.1111/rssc.12417   
6140   https://doi.org/10.1053/j.seminhematol.2020.05...   
6154                   https://doi.org/10.1093/sq/quz014   
6287       https://doi.org/10.1016/s0038-1101(13)00240-2   
6288           https://doi.org/10.1016/j.sse.2014.07.004   
6290            https://doi.org/10.1215/00382876-1162597   
6292            https://doi.org/10.1215/00382876-2345270   
6300    https://doi.org/10.1111/j.1540-6237.2010.00686.x   
6301    https://doi.org/10.1111/j.1540-6237.2011.00776.x   
6303                  https://doi.org/10.1111/ssqu.12006   
6310              https://doi.org/10.1002/star.201000005   
6314    https://doi.org/10.1111/j.1467-9582.2012.01187.x   
7037               https://doi.org/10.1353/sub.2012.0010   
7038               https://doi.org/10.1353/sub.2014.0026   
7287           https://doi.org/10.1017/s0075426920000300   
7408           https://doi.org/10.1017/s0080440114000073   
74

Initial analysis shows that the 3-10K loop was a success despite the long warning/error messages during execution.

Only 58 got Zero-Moscow error, which will be fixed below.

The date success rate seems to be again around 50%


In [24]:
Scihub_df.to_csv("15K_done_Scihub_df.csv")

In [55]:
%store Scihub_df

Stored 'Scihub_df' (DataFrame)


In [13]:
sh = SciHub()

In [9]:
Scihub_df.loc[Scihub_df.Scihub_results=="cant_read_pdf (Zero-Moscow)", "Scihub_results"] = Scihub_df.loc[Scihub_df.Scihub_results=="cant_read_pdf (Zero-Moscow)", "doi"].map(sh.get_dates)

In [11]:
Scihub_df.loc[Scihub_df.Scihub_results=="cant_read_pdf (Zero-Moscow)", :]

doi  \
5698  10.1097/psy.0000000000000377   
7037         10.1353/sub.2012.0010   
7038         10.1353/sub.2014.0026   
9105  10.1097/sap.0000000000000959   

                                           doi_url  \
5698  https://doi.org/10.1097/psy.0000000000000377   
7037         https://doi.org/10.1353/sub.2012.0010   
7038         https://doi.org/10.1353/sub.2014.0026   
9105  https://doi.org/10.1097/sap.0000000000000959   

                                                  title            genre  \
5698  Low-Grade Inflammation and Ambulatory Cortisol...  journal-article   
7037    How To (Un)Globe the Earth in Four Easy Lessons  journal-article   
7038  Sovereignty's Ontological Indecision: Derrida ...  journal-article   
9105            Pediatric Hand Surgery in Global Health  journal-article   

     published_date  year               journal_name        journal_issns  \
5698     2017-02-01  2017     Psychosomatic Medicine  1534-7796,0033-3174   
7037     2012-01-01  2012                  SubStance  0049-2426,1527-2095   
7038     2014-01-01  2014                  SubStance  0049-2426,1527-2095   
9105     2017-02-01  2017  Annals of Plastic Surgery            0148-7043   

     journal_issn_l  journal_is_oa  journal_is_in_doaj  \
5698      0033-3174          False               False   
7037      0049-2426          False               False   
7038      0049-2426          False               False   
9105      0148-7043          False               False   

                                      publisher  is_oa oa_status  \
5698  Ovid Technologies (Wolters Kluwer Health)   True     green   
7037                               Project Muse  False    closed   
7038                               Project Muse  False    closed   
9105  Ovid Technologies (Wolters Kluwer Health)  False    closed   

                                           oa_locations  \
5698  [{'updated': None, 'url': 'https://europepmc.o...   
7037                                                 []   
7038                                                 []   
9105                                                 []   

                         updated  \
5698  2021-11-27T10:44:18.234423   
7037  2021-01-18T22:49:18.155121   
7038  2021-01-15T02:54:35.260580   
9105  2021-01-19T10:47:17.809953   

                                              z_authors Unpy_filter  \
5698  [{'given': 'Hannah M.C.', 'family': 'Schreier'...      Scihub   
7037  [{'given': 'R.', 'family': 'Ghosh', 'sequence'...      Scihub   
7038  [{'given': 'M.', 'family': 'Chrulew', 'sequenc...      Scihub   
9105  [{'given': 'Karen Y.', 'family': 'Chung', 'seq...      Scihub   

                   Scihub_results  
5698  cant_read_pdf (Zero-Moscow)  
7037  cant_read_pdf (Zero-Moscow)  
7038  cant_read_pdf (Zero-Moscow)  
9105  cant_read_pdf (Zero-Moscow)

All but 4 Zero-Moscow's are fixed. However, it is seen that these 4 PDFs have a different font which is not-readable by both pdfminer and PyPdf2. These will not be fixed and will be left as an error. However, this will require an update on SciHub class.

Likewise, during this run it is seen that the errors raised during the mapping process are not exactly working as intended. Although the "retry"s fix the problem for the next DOI, the article at hand is a "lost cause" as all the variables for that specific loop stay the same. So, the error & retry processes need to be changed entirely, so the problematic DOI at hand is also fixed, and no retried unnecessarily.

scihub_upgraded

## Third Loop (15K - 20K)

In [9]:
for k in range(5):

    sh = SciHub()
    
    start = (k+15)*1000
    end = (k+16)*1000 
    doi_slice = Scihub_df[start:end].index.tolist()

    now = datetime.now()
    print(f"Starting loop {k+1}, start time: {now}")
    Scihub_df.loc[doi_slice,"Scihub_results"] = Scihub_df.loc[doi_slice,"doi"].map(sh.get_dates)
    now = datetime.now()
    print(f"{k+1} loop ended, endtime: {now}")
   

Starting loop 1, start time: 2022-01-31 16:03:37.656035


1 loop ended, endtime: 2022-01-31 16:53:23.746616
Starting loop 2, start time: 2022-01-31 16:53:23.750616


2 loop ended, endtime: 2022-01-31 17:43:56.142533
Starting loop 3, start time: 2022-01-31 17:43:56.147035


3 loop ended, endtime: 2022-01-31 18:36:04.841939
Starting loop 4, start time: 2022-01-31 18:36:04.845436


4 loop ended, endtime: 2022-01-31 19:27:08.726726
Starting loop 5, start time: 2022-01-31 19:27:08.730726


5 loop ended, endtime: 2022-01-31 20:16:40.794154


In [11]:
Scihub_df[15000:20000]["Scihub_results"].map(type).value_counts()

<class 'list'>    4856
<class 'str'>      144
Name: Scihub_results, dtype: int64

In [12]:
Scihub_df[15000:20000][Scihub_df[15000:20000]["Scihub_results"].map(type)==list]["Scihub_results"].map(len).value_counts()

3    2634
2    2222
Name: Scihub_results, dtype: int64

In [13]:
Scihub_df.to_csv("20K_Scihub_df.csv")

In [14]:
%store Scihub_df

Stored 'Scihub_df' (DataFrame)


First 20K of complete_one Scihub articles are completed. The rest 25K will be gathered from Çeto & Summan. In the meantime, run a loop for the failed direct url articles.

In [9]:
%store -r Unpy_one_fin

In [10]:
Unpy_suc_df = Unpy_one_fin.loc[Unpy_one_fin.Unpy_results.notna(),:].copy()

In [11]:
Unpy_suc_df.shape

(5966, 22)

In [13]:
Unpy_suc_df.to_csv("Unpy_suc_run_v2.csv")

In [3]:
Unpy_one_fin.Unpy_results.map(type).value_counts()

<class 'NoneType'>    11461
<class 'list'>         5966
Name: Unpy_results, dtype: int64

In [23]:
Unpy_fail_df = Unpy_one_fin.loc[Unpy_one_fin.Unpy_results.isna(),:].copy()

In [25]:
for k in range(12):

    sh = SciHub()
    
    start = (k)*1000
    end = (k+1)*1000

    if k == 11:
        doi_slice = Unpy_fail_df[start:].index.tolist()
    else:
        doi_slice = Unpy_fail_df[start:end].index.tolist()

    now = datetime.now()
    print(f"Starting loop {k+1}, start time: {now}")
    Unpy_fail_df.loc[doi_slice,"Scihub_results"] = Unpy_fail_df.loc[doi_slice,"doi"].map(sh.get_dates)
    now = datetime.now()
    print(f"{k+1} loop ended, endtime: {now}")
   

Starting loop 1, start time: 2022-01-31 23:46:03.984852


1 loop ended, endtime: 2022-02-01 00:46:52.189736
Starting loop 2, start time: 2022-02-01 00:46:52.195237


2 loop ended, endtime: 2022-02-01 01:53:49.334478
Starting loop 3, start time: 2022-02-01 01:53:49.339978


KeyboardInterrupt: 

In [27]:
Unpy_fail_df[:2000].Scihub_results.map(type).value_counts()

<class 'list'>    1705
<class 'str'>      295
Name: Scihub_results, dtype: int64

In [33]:
Unpy_fail_df[:2000][Unpy_fail_df[:2000].Scihub_results.map(type)== str]["Scihub_results"].value_counts()

article_not_in Scihub    288
cant_read_pdf              7
Name: Scihub_results, dtype: int64

Due to an error in domain list (sci-hub.ru), need to restart the loop from [2000:]


In [6]:
%store -r Unpy_fail_df

In [8]:
for k in range(2,12):

    sh = SciHub()
    
    start = (k)*1000
    end = (k+1)*1000

    if k == 11:
        doi_slice = Unpy_fail_df[start:].index.tolist()
    else:
        doi_slice = Unpy_fail_df[start:end].index.tolist()

    now = datetime.now()
    print(f"Starting loop {k+1}, start time: {now}")
    Unpy_fail_df.loc[doi_slice,"Scihub_results"] = Unpy_fail_df.loc[doi_slice,"doi"].map(sh.get_dates)
    now = datetime.now()
    print(f"{k+1} loop ended, endtime: {now}")
   

Starting loop 3, start time: 2022-02-01 09:24:35.486427


3 loop ended, endtime: 2022-02-01 10:19:10.040015
Starting loop 4, start time: 2022-02-01 10:19:10.044514


4 loop ended, endtime: 2022-02-01 11:28:23.920809
Starting loop 5, start time: 2022-02-01 11:28:23.923310


5 loop ended, endtime: 2022-02-01 12:14:37.506195
Starting loop 6, start time: 2022-02-01 12:14:37.510196


6 loop ended, endtime: 2022-02-01 13:06:13.020522
Starting loop 7, start time: 2022-02-01 13:06:13.026022


7 loop ended, endtime: 2022-02-01 13:58:19.051415
Starting loop 8, start time: 2022-02-01 13:58:19.055916


8 loop ended, endtime: 2022-02-01 14:57:14.495951
Starting loop 9, start time: 2022-02-01 14:57:14.500451


9 loop ended, endtime: 2022-02-01 16:09:30.152721
Starting loop 10, start time: 2022-02-01 16:09:30.154222


10 loop ended, endtime: 2022-02-01 17:37:36.663456
Starting loop 11, start time: 2022-02-01 17:37:36.667455


11 loop ended, endtime: 2022-02-01 18:44:29.498909
Starting loop 12, start time: 2022-02-01 18:44:29.500909


12 loop ended, endtime: 2022-02-01 19:30:32.901064


In [10]:
Unpy_fail_df.Scihub_results.map(type).value_counts()

<class 'list'>    10205
<class 'str'>      1256
Name: Scihub_results, dtype: int64

In [11]:
Unpy_fail_df[Unpy_fail_df.Scihub_results.map(type)==str]["Scihub_results"].value_counts()

article_not_in Scihub    1211
cant_read_pdf              41
direct_url_error            4
Name: Scihub_results, dtype: int64

In [14]:
Unpy_fail_df[Unpy_fail_df.Scihub_results.map(type)==list]["Scihub_results"].map(len).value_counts(normalize=True)

2    0.64831
3    0.35169
Name: Scihub_results, dtype: float64

In [17]:
Unpy_fail_df[Unpy_fail_df.Scihub_results=="article_not_in Scihub"]

doi  \
21            10.1590/0001-3675202020180560   
33               10.2308/accr.2010.85.2.749   
69             10.4310/acta.2017.v218.n1.a1   
70             10.4310/acta.2018.v221.n1.a5   
71             10.4310/acta.2019.v222.n2.a1   
...                                     ...   
64001               10.28991/esj-2020-01246   
64002               10.32609/j.ruje.6.50824   
64003  10.22162/2619-0990-2019-45-5-855-871   
64004  10.22162/2619-0990-2020-49-3-779-789   
64026  10.22363/2687-0088-2020-24-2-366-385   

                                                 doi_url  \
21         https://doi.org/10.1590/0001-3675202020180560   
33            https://doi.org/10.2308/accr.2010.85.2.749   
69          https://doi.org/10.4310/acta.2017.v218.n1.a1   
70          https://doi.org/10.4310/acta.2018.v221.n1.a5   
71          https://doi.org/10.4310/acta.2019.v222.n2.a1   
...                                                  ...   
64001            https://doi.org/10.28991/esj-2020-01246   
64002            https://doi.org/10.32609/j.ruje.6.50824   
64003  https://doi.org/10.22162/2619-0990-2019-45-5-8...   
64004  https://doi.org/10.22162/2619-0990-2020-49-3-7...   
64026  https://doi.org/10.22363/2687-0088-2020-24-2-3...   

                                                   title            genre  \
21     Size at onset of sexual maturity in Macrobrach...  journal-article   
33                Editorial Policy and Style Information  journal-article   
69                  Tits geometry and positive curvature  journal-article   
70     Isoperimetric characterization of upper curvat...  journal-article   
71       Bogoliubov theory in the Gross–Pitaevskii limit  journal-article   
...                                                  ...              ...   
64001  Analysis of Host Resources Utilization by Open...  journal-article   
64002  Prospects of the Chinese market for Russian ag...  journal-article   
64003  Импортная поливная керамика Маджара как источн...  journal-article   
64004  Мотив «богатырский поединок (сражение)» и его ...  journal-article   
64026  Multimodal IT marketing discourse: An integrat...  journal-article   

      published_date  year                              journal_name  \
21        2020-01-01  2020  Anais da Academia Brasileira de Ciências   
33        2010-03-01  2010                     The Accounting Review   
69        2017-01-01  2017                          Acta Mathematica   
70        2018-01-01  2018                          Acta Mathematica   
71        2019-01-01  2019                          Acta Mathematica   
...              ...   ...                                       ...   
64001     2020-12-01  2020                  Emerging Science Journal   
64002     2020-03-25  2020              Russian Journal of Economics   
64003     2019-12-11  2019                          Oriental Studies   
64004     2020-12-24  2020                          Oriental Studies   
64026     2020-12-15  2020            Russian Journal of Linguistics   

             journal_issns journal_issn_l  journal_is_oa  ...  oa_status  \
21     1678-2690,0001-3765      0001-3765           True  ...       gold   
33     0001-4826,1558-7967      0001-4826          False  ...     bronze   
69     0001-5962,1871-2509      0001-5962           True  ...       gold   
70     0001-5962,1871-2509      0001-5962           True  ...       gold   
71     0001-5962,1871-2509      0001-5962           True  ...       gold   
...                    ...            ...            ...  ...        ...   
64001            2610-9182      2610-9182           True  ...       gold   
64002  2405-4739,2618-7213      2405-4739           True  ...       gold   
64003  2619-0990,2619-1008      2619-0990           True  ...       gold   
64004  2619-0990,2619-1008      2619-0990           True  ...       gold   
64026  2686-8024,2687-0088      2686-8024           True  ...       gold   

                                            oa_loca

In [15]:
Unpy_fail_df.to_csv("Unpy_scihub_run.csv")

In [16]:
%store Unpy_fail_df

Stored 'Unpy_fail_df' (DataFrame)


MANUAL GET_DATES TESTING:

In [37]:
base_url = "https://sci-hub.se/"
doi = '10.1097/psy.0000000000000377'
doi2 = "10.1353/sub.2012.0010"
doi4 = "10.1097/sap.0000000000000959"

total_url = base_url + doi4

sh_req = requests.get(total_url)
sh_soup = BeautifulSoup(sh_req.content, "html.parser")
iframe = sh_soup.find("embed", {"id": "pdf"})["src"]


In [38]:
iframe

'https://twin.sci-hub.se/6624/192e4f171d0db2136e32b26a6dc80550/chung2017.pdf#navpanes=0&view=FitH'

In [39]:
pdf_req = requests.get(iframe)


In [40]:
pdf_io = io.BytesIO(pdf_req.content)


In [41]:
pdf_read = extract_text(pdf_io,page_numbers=[]).replace("\n", "")


RecursionError: maximum recursion depth exceeded

In [34]:
pdf2_read = pdf2.PdfFileReader(pdf_io)

In [36]:
pdf2_read.getPage(1)

{'/TrimBox': [0, 0, 432, 648],
 '/Resources': {'/Font': {'/C2_0': {'/Encoding': '/Identity-H',
    '/ToUnicode': {'/Filter': '/FlateDecode'},
    '/Subtype': '/Type0',
    '/Type': '/Font',
    '/DescendantFonts': [IndirectObject(57, 0)],
    '/BaseFont': '/FGDIQC+Palatino-Roman'},
   '/TT4': {'/Encoding': '/WinAnsiEncoding',
    '/ToUnicode': {'/Filter': '/FlateDecode'},
    '/Widths': [250,
     0,
     0,
     0,
     0,
     0,
     0,
     0,
     0,
     0,
     0,
     0,
     250,
     333,
     250,
     0,
     0,
     0,
     0,
     0,
     0,
     0,
     0,
     0,
     0,
     0,
     0,
     0,
     0,
     0,
     0,
     0,
     0,
     722,
     611,
     667,
     778,
     611,
     556,
     722,
     0,
     333,
     0,
     0,
     556,
     944,
     778,
     0,
     611,
     0,
     667,
     556,
     611,
     778,
     0,
     944,
     0,
     0,
     0,
     0,
     0,
     0,
     0,
     0,
     0,
     444,
     0,
     407,
     500,
     389,
    

In [19]:
sh.get_dates("10.1016/j.jinsphys.2011.12.018")

['https://sci.bban.top/pdf/10.1016/j.jinsphys.2011.12.018.pdf#view=FitH',
 'no_date_found']

TESTING AREA:

In [14]:
from retry import retry

In [15]:
@retry((TypeError), tries=2, delay=10)
@retry((ValueError), tries=2, delay=15)
def bokbok(inpo):
    if inpo == 1:
        raise TypeError("1")
    elif inpo == 2:
        raise ValueError("2")

In [17]:
bokbok(1)


TypeError: 1

In [18]:
sh.get_dates("askjdhsakd")

ReadTimeout: HTTPSConnectionPool(host='sci-hub.wf', port=443): Read timed out. (read timeout=15)

In [37]:
Scihub_df[:2000]["Scihub_results"].map(type).value_counts()

<class 'list'>    1870
<class 'str'>      130
Name: Scihub_results, dtype: int64

In [40]:
Scihub_df[:2000][Scihub_df[:2000]["Scihub_results"].map(type)==str]["Scihub_results"].value_counts()

article_not_in Scihub    130
Name: Scihub_results, dtype: int64

In [38]:
first_run_df[:2000]["Scihub_results"].map(type).value_counts()

<class 'list'>    1393
<class 'str'>      607
Name: Scihub_results, dtype: int64

In [67]:
first_run_df[:2000][first_run_df[:2000]["Scihub_results"].map(type)==str]["Scihub_results"].value_counts()

cant_open_pdf            493
article_not_in Scihub    114
Name: Scihub_results, dtype: int64

In [76]:
run2_niS = Scihub_df[:2000][Scihub_df[:2000]["Scihub_results"] == "article_not_in Scihub"]

In [77]:
run1_niS = first_run_df[:2000][first_run_df[:2000]["Scihub_results"] == "article_not_in Scihub"]

In [84]:
run2_niS.drop(run1_niS.index.tolist())

doi  \
98    10.1080/00016357.2020.1859132   
170       10.1017/s0001972020000790   
360         10.5344/ajev.2020.19050   
460                  10.1086/710566   
487            10.2138/am-2017-6003   
798                 10.1145/3366018   
799                 10.1145/3398745   
1059         10.1093/aesthj/ayaa043   
1155      10.1017/s000712342000006x   
1177        10.1111/1468-4446.12739   
1275         10.4153/cjm-2010-073-8   
1470      10.1017/s0009838820000671   
1559                 10.1086/709687   
1765       10.1177/0011392117737820   
2107            10.1093/ehr/ceaa223   
2497     10.13109/glot.2020.96.1.82   

                                            doi_url  \
98    https://doi.org/10.1080/00016357.2020.1859132   
170       https://doi.org/10.1017/s0001972020000790   
360         https://doi.org/10.5344/ajev.2020.19050   
460                  https://doi.org/10.1086/710566   
487            https://doi.org/10.2138/am-2017-6003   
798                 https://doi.org/10.1145/3366018   
799                 https://doi.org/10.1145/3398745   
1059         https://doi.org/10.1093/aesthj/ayaa043   
1155      https://doi.org/10.1017/s000712342000006x   
1177        https://doi.org/10.1111/1468-4446.12739   
1275         https://doi.org/10.4153/cjm-2010-073-8   
1470      https://doi.org/10.1017/s0009838820000671   
1559                 https://doi.org/10.1086/709687   
1765       https://doi.org/10.1177/0011392117737820   
2107            https://doi.org/10.1093/ehr/ceaa223   
2497     https://doi.org/10.13109/glot.2020.96.1.82   

                                                  title            genre  \
98    Association between periodontal disease and in...  journal-article   
170   Devaka Premawardhana, Faith in Flux: Pentecost...  journal-article   
360   Meteorological-Based Modeling of δ18O Values f...  journal-article   
460   Syndicate Women: Gender and Networks in Chicag...  journal-article   
487   Study on structure variations of incommensurat...  journal-article   
798   On the Complexity of Cache Analysis for Differ...  journal-article   
799   A Simple and Approximately Optimal Mechanism f...  journal-article   
1059  White Negroes: When Cornrows Were in Vogue … a...  journal-article   
1155  An ‘Institution-First’ Conception of Public In...  journal-article   
1177  What do stress tests test? Experimentation, de...  journal-article   
1275                        On Vojta's 1 + Ɛ Conjecture  journal-article   
1470  VISUALIZING THE SHIELD OF ACHILLES: APPROACHIN...  journal-article   
1559  Midnight Traveler directed by Hassan Fazili. P...  journal-article   
1765  Unravelling ambivalence: A field-theoretical a...  journal-article   
2107  Birthright Citizens: A History of Race and Rig...  journal-article   
2497  The past-oriented aorist infinitive in Post-Cl...  journal-article   

     published_date  year                                 journal_name  \
98       2020-12-28  2020               Acta Odontologica Scandinavica   
170      2020-11-01  2020                                       Africa   
360      2020-03-13  2020  American Journal of Enology and Viticulture   
460      2020-11-01  2020                American Journal of Sociology   
487      2017-06-01  2017                        American Mineralogist   
798      2019-12-06  2019                           Journal of the ACM   
799      2020-08-13  2020                           Journal of the ACM   
1059     2020-12-17  2020            The British Journal of Aesthetics   
1155     2020-11-17  2020         British Journal of Political Science   
1177     2020-02-12  2020             The British Journal of Sociology   
1275     2011-02-01  2011              Canadian Journal of Mathematics   
1470     2020-12-01  2020                      The Classical Quarterly   
1559     2020-08-01  2020                 Comparative Education Review   
1765     2017-11-20  2017                            Current Sociology   
2107     2020-10-01  2020    

In [118]:
sh = SciHub()

sh.get_dates("10.1093/ehr/ceaa223")

TypeError: unsupported operand type(s) for +: 'NoneType' and 'str'

In [128]:
base_url = 'https://sci-hub.ru/'
doi = "10.1093/ehr/ceaa223"

total_url = base_url + doi

sh_req = requests.get(total_url)
sh_soup = BeautifulSoup(sh_req.content, "html.parser")
iframe = sh_soup.find("iframe", {"id": "pdf"})["src"]



TypeError: 'NoneType' object is not callable

In [131]:
iframe = sh_soup.find("iframe", {"id": "pdf"})["src"]

TypeError: 'NoneType' object is not subscriptable

In [133]:

iframe = sh_soup.find("embed", {"id": "pdf"})["src"]

In [134]:
iframe

'//sci-hub.ru/downloads/2021-05-26/3553/oup-accepted-manuscript-2020.pdf#navpanes=0&view=FitH'

In [ ]:
if iframe.startswith("//"):
    iframe = 'http:' + iframe

pdf_req = requests.get(iframe)
pdf_io = io.BytesIO(pdf_req.content)
pdf_read = extract_text(pdf_io,page_numbers=[]).replace("\n", "")

print(pdf_read[:10])